private/pkg/manifest/manifest.go

```diff
@@ -0,0 +1,269 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Manifests are a list of paths and their hash digests, canonically ordered by
+// path in increasing lexographical order. Manifests are encoded as:
+//
+//	<hash type>:<digest>[SP][SP]<path>[LF]
+//
+// "shake256" is the only supported hash type. The digest is 64 bytes of hex
+// encoded output of SHAKE256. See golang.org/x/crypto/sha3 and FIPS 202 for
+// details on the SHAKE hash.
+package manifest
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/bufbuild/buf/private/pkg/storage"
+	"go.uber.org/multierr"
+	"golang.org/x/crypto/sha3"
+)
+
+var errPartial error = errors.New("partial record")
+
+const shake256Name = "shake256"
+
+type Digest struct {
+	dtype  string
+	digest []byte
+	hexstr string
+}
+
+func NewDigestFromBytes(dtype string, digest []byte) *Digest {
```

Comment:
Since the other `New*` methods return a possible error, I wonder if we should do the same and do basic validation here (dtype/digest non-empty?)

Reply:
This would be the wrong place to check for supported types (`Digest` doesn't care what you stick in it), but I can add the check in `Manifest.addDigest`.

Reply:
If no validations are done here, and we'll always succeed returning a Digest, then I suggest returning a value.
```suggestion
func NewDigestFromBytes(dtype string, digest []byte) Digest {
```

Reply:
Doh, I missed this in the commits I just pushed. Hold on and I'll make this change (ret by value works for me).

Reply:
On second pass, returning a pointer makes for better intuitive use with the pointer receivers. Lines like this will break:

```go
fmt.Printf("%s", digest)
```

Stringing digests would need to be passed as `&digest`.

There's been some discussion on this topic in the general style guide (pointer all the things), so I'm going to stick to returning and using pointers for now.